In [ ]:
import pandas as pd

In [ ]:
# path the dataset
dataset_path = 'uiuc-gpa-dataset.csv'

df_head = pd.read_csv(dataset_path)
df_columns = df_head.columns.tolist()
print("current columns:", df_columns)
print("\n the first 5 rows:")
df_head.head()

current columns: ['Year', 'Term', 'YearTerm', 'Subject', 'Number', 'Course Title', 'Sched Type', 'A+', 'A', 'A-', 'B+', 'B', 'B-', 'C+', 'C', 'C-', 'D+', 'D', 'D-', 'F', 'W', 'Primary Instructor']

 the first 5 rows:


Year    Term YearTerm Subject  Number                  Course Title  \
0  2023  Spring  2023-sp     AAS     100  Intro Asian American Studies   
1  2023  Spring  2023-sp     AAS     100  Intro Asian American Studies   
2  2023  Spring  2023-sp     AAS     100  Intro Asian American Studies   
3  2023  Spring  2023-sp     AAS     200          U.S. Race and Empire   
4  2023  Spring  2023-sp     AAS     215  US Citizenship Comparatively   

  Sched Type  A+   A  A-  ...  B-  C+  C  C-  D+  D  D-  F  W  \
0        DIS   0  11   5  ...   0   0  0   0   0  0   0  1  0   
1        DIS   0  17   2  ...   1   0  0   0   0  0   0  0  1   
2        DIS   0  13   2  ...   2   0  0   1   0  0   0  1  0   
3        LCD   6  15   5  ...   0   0  0   0   0  1   0  1  0   
4        LCD  16  12   2  ...   1   0  0   0   0  0   0  0  0   

   Primary Instructor  
0       Shin, Jeongsu  
1       Shin, Jeongsu  
2      Lee, Sabrina Y  
3      Sawada, Emilia  
4        Kwon, Soo Ah  

[5 rows x 22 columns]

In [ ]:
grade_columns = ['A+', 'A', 'A-', 'B+', 'B', 'B-', 'C+', 'C', 'C-', 'D+', 'D', 'D-', 'F', 'W']

df_head['Total_Students'] = df_head[grade_columns].sum(axis=1)

for grade in grade_columns:
    df_head[grade] = df_head[grade] / df_head['Total_Students']

df_head.head()

Year    Term YearTerm Subject  Number                  Course Title  \
0  2023  Spring  2023-sp     AAS     100  Intro Asian American Studies   
1  2023  Spring  2023-sp     AAS     100  Intro Asian American Studies   
2  2023  Spring  2023-sp     AAS     100  Intro Asian American Studies   
3  2023  Spring  2023-sp     AAS     200          U.S. Race and Empire   
4  2023  Spring  2023-sp     AAS     215  US Citizenship Comparatively   

  Sched Type        A+         A        A-  ...   C+    C        C-   D+  \
0        DIS  0.000000  0.500000  0.227273  ...  0.0  0.0  0.000000  0.0   
1        DIS  0.000000  0.708333  0.083333  ...  0.0  0.0  0.000000  0.0   
2        DIS  0.000000  0.619048  0.095238  ...  0.0  0.0  0.047619  0.0   
3        LCD  0.181818  0.454545  0.151515  ...  0.0  0.0  0.000000  0.0   
4        LCD  0.484848  0.363636  0.060606  ...  0.0  0.0  0.000000  0.0   

          D   D-         F         W  Primary Instructor  Total_Students  
0  0.000000  0.0  0.045455  0.000000       Shin, Jeongsu              22  
1  0.000000  0.0  0.000000  0.041667       Shin, Jeongsu              24  
2  0.000000  0.0  0.047619  0.000000      Lee, Sabrina Y              21  
3  0.030303  0.0  0.030303  0.000000      Sawada, Emilia              33  
4  0.000000  0.0  0.000000  0.000000        Kwon, Soo Ah              33  

[5 rows x 23 columns]

In [ ]:
# check the number of unique instructors
unique_instructors = df_head['Primary Instructor'].unique()
print("number of unique instructors:", len(unique_instructors))

# number of data points
print("number of data points:", len(df_head))

number of unique instructors: 9313
number of data points: 69069


we have 69069 datapoints (training + validation + test) in total, but there are 9313 unique primary instructors. It means that it is likely to lead the model to too high dimension to deal with. Therefore, we would like to drop the primary instructor feature.

In [ ]:
# remove columns that are uncessary, not useful
df_head = df_head.drop(columns=['Primary Instructor', 'Total_Students', 'YearTerm'])
df_head.head()

Year    Term Subject  Number                  Course Title Sched Type  \
0  2023  Spring     AAS     100  Intro Asian American Studies        DIS   
1  2023  Spring     AAS     100  Intro Asian American Studies        DIS   
2  2023  Spring     AAS     100  Intro Asian American Studies        DIS   
3  2023  Spring     AAS     200          U.S. Race and Empire        LCD   
4  2023  Spring     AAS     215  US Citizenship Comparatively        LCD   

         A+         A        A-        B+         B        B-   C+    C  \
0  0.000000  0.500000  0.227273  0.045455  0.181818  0.000000  0.0  0.0   
1  0.000000  0.708333  0.083333  0.041667  0.083333  0.041667  0.0  0.0   
2  0.000000  0.619048  0.095238  0.095238  0.000000  0.095238  0.0  0.0   
3  0.181818  0.454545  0.151515  0.060606  0.090909  0.000000  0.0  0.0   
4  0.484848  0.363636  0.060606  0.030303  0.030303  0.030303  0.0  0.0   

         C-   D+         D   D-         F         W  
0  0.000000  0.0  0.000000  0.0  0.045455  0.000000  
1  0.000000  0.0  0.000000  0.0  0.000000  0.041667  
2  0.047619  0.0  0.000000  0.0  0.047619  0.000000  
3  0.000000  0.0  0.030303  0.0  0.030303  0.000000  
4  0.000000  0.0  0.000000  0.0  0.000000  0.000000

In [ ]:
# check the available features at this point
df_head.columns.tolist()

['Year',
 'Term',
 'Subject',
 'Number',
 'Course Title',
 'Sched Type',
 'A+',
 'A',
 'A-',
 'B+',
 'B',
 'B-',
 'C+',
 'C',
 'C-',
 'D+',
 'D',
 'D-',
 'F',
 'W']

In [ ]:
categorical_columns = ['Year', 'Term', 'Subject', 'Sched Type', 'Number', 'Course Title']

# check the unique features for each

for column in categorical_columns:
    unique_values = df_head[column].unique()
    print(f"unique values for {column}: {len(unique_values)}")

unique values for Year: 14
unique values for Term: 4
unique values for Subject: 172
unique values for Sched Type: 19
unique values for Number: 556
unique values for Course Title: 5813


In [ ]:
dataset = df_head.copy()
dataset = dataset.dropna(subset=categorical_columns)
encoding_dict = {}

# Assign a unique integer for each unique value in categorical features
for column in categorical_columns:
    # Extract unique values directly from df_head, not the dataset copy
    unique_values = df_head[column].dropna().astype(str).unique()
    unique_values = sorted(unique_values)  # Sort alphabetically

    # Create mapping
    value_to_int_mapping = {value: idx for idx, value in enumerate(unique_values, start=0)}
    encoding_dict[column] = value_to_int_mapping

    # Map values and fill NaNs for missing mappings
    dataset[column] = dataset[column].astype(str).map(value_to_int_mapping).astype(int)

In [ ]:
print(dataset['Year'].unique())

print(df_head['Year'].unique())

[13 12 11 10  9  8  6  5  4  3  2  1  0]
[2023 2022 2021 2020 2019 2018 2017 2016 2015 2014 2013 2012 2011 2010]


In [ ]:
# check if there is any NaN
dataset.isna().sum()

Year            0
Term            0
Subject         0
Number          0
Course Title    0
Sched Type      0
A+              0
A               0
A-              0
B+              0
B               0
B-              0
C+              0
C               0
C-              0
D+              0
D               0
D-              0
F               0
W               0
dtype: int64

In [ ]:
dataset.head()

Year  Term  Subject  Number  Course Title  Sched Type        A+         A  \
0    13     1        0       0          2893           1  0.000000  0.500000   
1    13     1        0       0          2893           1  0.000000  0.708333   
2    13     1        0       0          2893           1  0.000000  0.619048   
3    13     1        0      86          5550           6  0.181818  0.454545   
4    13     1        0     101          5556           6  0.484848  0.363636   

         A-        B+         B        B-   C+    C        C-   D+         D  \
0  0.227273  0.045455  0.181818  0.000000  0.0  0.0  0.000000  0.0  0.000000   
1  0.083333  0.041667  0.083333  0.041667  0.0  0.0  0.000000  0.0  0.000000   
2  0.095238  0.095238  0.000000  0.095238  0.0  0.0  0.047619  0.0  0.000000   
3  0.151515  0.060606  0.090909  0.000000  0.0  0.0  0.000000  0.0  0.030303   
4  0.060606  0.030303  0.030303  0.030303  0.0  0.0  0.000000  0.0  0.000000   

    D-         F         W  
0  0.0  0.045455  0.000000  
1  0.0  0.000000  0.041667  
2  0.0  0.047619  0.000000  
3  0.0  0.030303  0.000000  
4  0.0  0.000000  0.000000

# Making debugging dataset (here we use 5% of the original dataset) and working dataset (original dataset with data cleaning)

In [ ]:
debugging_dataset = dataset.sample(frac=0.05, random_state=42)
working_dataset = dataset.copy()

print("debugging dataset shape:", debugging_dataset.shape)
print("working dataset shape:", working_dataset.shape)

debugging dataset shape: (2951, 20)
working dataset shape: (59016, 20)


# Converting the datasets to Pickles

In [ ]:
debugging_dataset.to_pickle('debugging_dataset.pkl')
working_dataset.to_pickle('working_dataset.pkl')